service/turn.proxy/src/main/java/org/kaazing/gateway/service/turn/proxy/stun/StunFrameEncoder.java

```diff
@@ -0,0 +1,151 @@
+
+import static org.kaazing.gateway.service.turn.proxy.stun.StunProxyMessage.attributePaddedLength;
+import static org.kaazing.gateway.util.turn.TurnUtils.HMAC_SHA_1;
+
+
+/*
+   https://tools.ietf.org/html/rfc5389#section-15.4
+
+   Based on the rules above, the hash used to construct MESSAGE-
+   INTEGRITY includes the length field from the STUN message header.
+   Prior to performing the hash, the MESSAGE-INTEGRITY attribute MUST be
+   inserted into the message (with dummy content).  The length MUST then
+   be set to point to the length of the message up to, and including,
+   the MESSAGE-INTEGRITY attribute itself, but excluding any attributes
+   after it.  Once the computation is performed, the value of the
+   MESSAGE-INTEGRITY attribute can be filled in, and the value of the
+   length in the STUN header can be set to its correct value -- the
+   length of the entire message.  Similarly, when validating the
+   MESSAGE-INTEGRITY, the length field should be adjusted to point to
+   the end of the MESSAGE-INTEGRITY attribute prior to calculating the
+   HMAC.  Such adjustment is necessary when attributes, such as
+   FINGERPRINT, appear after MESSAGE-INTEGRITY.
+ */
+
+public class StunFrameEncoder extends ProtocolEncoderAdapter {
+
+    private final IoBufferAllocatorEx<?> allocator;
+    private final ConcurrentMap<String, String> currentTransactions;
+    private final Key sharedSecret;
+    private final String keyAlgorithm;
+
+    private static final Logger LOGGER = LoggerFactory.getLogger("service.turn.proxy");
+
+
+    public StunFrameEncoder(IoBufferAllocatorEx<?> aloc, ConcurrentMap<String, String> currentTrx, Key ss, String keyAlg) {
+        this.allocator = aloc;
+        this.currentTransactions = currentTrx;
+        this.sharedSecret = ss;
+        this.keyAlgorithm = keyAlg;
+    }
+
+    @Override
+    public void encode(IoSession session, Object message, ProtocolEncoderOutput out) throws Exception {
+        if (!(message instanceof StunProxyMessage)) {
+            // easiest way to avoid race condition where decoder is removed on the filter chain prior to encoder
+            out.write(message);
+        }
+        if (LOGGER.isTraceEnabled()) {
+            LOGGER.trace("Encoding STUN message: " + message);
+        }
+        StunProxyMessage stunMessage = (StunProxyMessage) message;
+        String username = null;
+        if (stunMessage.getMessageClass().equals(StunMessageClass.RESPONSE) ||
+            stunMessage.getMessageClass().equals(StunMessageClass.ERROR)) {
+            username = currentTransactions.remove(Base64.getEncoder().encodeToString(stunMessage.getTransactionId()));
+            LOGGER.trace(String.format("Removed username %s from transactions map", username));
+            if (stunMessage.isModified() && (username == null || sharedSecret ==null)) {
+                LOGGER.warn("Stun message is modified but MESSAGE-INTEGRITY attribute will not be recalculated.");
```

Comment:
%s/Stun message is modified but MESSAGE-INTEGRITY attribute will not be recalculated/STUN message is modified but MESSAGE-INTEGRITY attribute can not be recalculated because username and/or shared secret is not available/